### ISSUE TITLE
Empty fragment results when calculating DMET energy with democratic partitioning from multiple MPI ranks

### ISSUE THREAD

**Issue by maxbortone** [BoothGroup/Vayesta#165]
I'm trying to use the callback solver with MPI to solve multiple fragments simultaneously, but I'm running into the issue where the fragment results are not being properly communicated between MPI ranks when calculating the DMET energy with democratic partitioning. The minimal example belowe reproduces the error when executed on 4 MPI rank with `mpirun -n 4`.

The problem appears on [L39](https://github.com/BoothGroup/Vayesta/blob/7023754519871a317f0b1bbaaefd7caf61faa2dc/vayesta/core/qemb/rdm.py#L39) of `vayesta/core/qemb/rdm.py`, where the fragment wavefunction is bein accessed to calculate the 1-RDM of the cluster. Since each fragment is on a separate MPI rank, not all results are available, leading to the following `AttributeError: 'NoneType' object has no attribute 'wf'`. 

```python
import numpy as np
import pyscf
import pyscf.gto
import pyscf.scf
import pyscf.fci
import vayesta
import vayesta.ewf
from vayesta.misc.molecules import ring
from vayesta import mpi

# User defined FCI solver - takes pyscf mf as input and returns RDMs
# The mf argment contains the hamiltonain in the orthonormal cluster basis
# Pyscf or other solvers may be used to solve the cluster problem and may return RDMs, CISD amplitudes or CCSD amplitudes
def solver(mf):
    h1e = mf.get_hcore()
    h2e = mf._eri
    norb = mf.mo_coeff.shape[-1]
    nelec = mf.mol.nelec
    energy, civec = pyscf.fci.direct_spin0.kernel(h1e, h2e, norb, nelec, conv_tol=1.e-14)
    dm1, dm2 = pyscf.fci.direct_spin0.make_rdm12(civec, norb, nelec)
    results = dict(dm1=dm1, dm2=dm2, converged=True)
    return results

natom = 4
mol = pyscf.gto.Mole()
mol.atom = ring("H", natom, 1.5)
mol.basis = "sto-3g"
mol.output = "pyscf.out"
mol.verbose = 5
mol.symmetry = True
mol.build()

# Hartree-Fock
mf = pyscf.scf.RHF(mol)
mf = mpi.scf(mf)
mf.kernel()

# Vayesta options
nfrag = 4
bath_opts = dict(bathtype="dmet")   

# Run vayesta with user defined solver
emb = vayesta.ewf.EWF(mf, solver="CALLBACK",  energy_functional='dm', bath_options=bath_opts, solver_options=dict(callback=solver))
# Set up fragments
with emb.iao_fragmentation() as f:
    # Add all atoms as separate fragments 
    f.add_all_atomic_fragments()
emb.kernel()

for x in emb.fragments:
    print(x.results)

# Calculate DMET energies
dmet_energy = emb.get_dmet_energy(part_cumulant=False, approx_cumulant=False)
dmet_energy_part_cumulant = emb.get_dmet_energy(part_cumulant=True, approx_cumulant=False)
dmet_energy_approx_cumulant = emb.get_dmet_energy(part_cumulant=True, approx_cumulant=True)

if mpi.is_master:
    print("Hartree-Fock energy          : %s"%mf.e_tot)
    print("DMET energy                  : %s"%dmet_energy)
    print("DMET energy   (part-cumulant): %s"%dmet_energy_part_cumulant)
    print("DMET energy (approx-cumulant): %s"%dmet_energy_approx_cumulant)
```

Probably something to add would be a function that broadcasts the fragment results across all MPI ranks, e.g.:
```python
def communicate_results(self):
    for x in emb.get_fragments(contributes=True):
        source = x.mpi_rank
        if source != mpi.rank:
            results = None
        else:
            results = x.results
        results = mpi.world.bcast(results, root=source)
        x.results = results
```

**Comment by basilib** [comment]
I had a look and it seems as though MPI is not implemented for democratically partitioned RDMs. I've implemented this in #166, now your example should be working and give correct results. I'll merge this soon but I need to make some additional changes for the global 2DM which isn't required for the energy.

**Comment by basilib** [comment]
Now merged into master.
